### PR TITLE
[FLINK-23090] Introduce RpcSystem abstraction

### PR DESCRIPTION
**Pull request by zentol**
Adds the `RpcSystem` interface which isolates flink-runtime from the AkkaUtils/AkkaRpcServiceUtils etc. . It serves as a factory for rpc services, as well as a crude facade for various static utilities. The interface is heavily inspired by the existing akka-related code; which isn't _that_ nice but fine for the time being because we don't intend to implement other RpcServices at this time.

Implementations of this interface (i.e., the AkkaRpcSystem) are loaded via `RpcSystem#load()` which currently just returns an AkkaRpcSystem, but Later we will load implementations of this interface with a `PluginLoader`.

### CODE REVIEW

**Review by rmetzger — COMMENTED** [review]
I made a pass over the changes. I have limited experience with the Akka / RPC system, but besides some cosmetic questions, the change seems fine.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java [hunk 1]
```diff
@@ -158,6 +158,10 @@ public static String getHostname(RpcService rpcService) {
                 : rpcServiceAddress;
     }
 
+    public static RpcSystem.ForkJoinExecutorConfiguration getTestForkJoinExecutorConfiguration() {
+        return new RpcSystem.ForkJoinExecutorConfiguration(1.0, 2, 4);
+    }
+
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcUtils.java:164]: hmm there should be more usages; I must've stopped porting them midway through.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistryConfiguration.java [hunk 2]
```diff
@@ -70,13 +70,19 @@ public long getQueryServiceMessageSizeLimit() {
     //  Static factory methods
     // ------------------------------------------------------------------------
 
+    @VisibleForTesting
+    public static MetricRegistryConfiguration fromConfiguration(Configuration configuration) {
+        return fromConfiguration(configuration, 10485760);
+    }
```
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistryConfiguration.java:76]: Test shouldn't have to worry about the maximum frame size; its a rather annoying implementation detail of the RPC system. I've refactored it into a test utility.

File: flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusReporterTest.java [hunk 3]
```diff
@@ -85,7 +85,7 @@
     public void setupReporter() {
         registry =
                 new MetricRegistryImpl(
-                        MetricRegistryConfiguration.defaultMetricRegistryConfiguration(),
+                        MetricRegistryTestUtils.defaultMetricRegistryConfiguration(),
```
Comment (tillrohrmann) [inline flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusReporterTest.java:88]: This is a bit unrelated to this change. But it looks to me that this test is more of an integration test which could go into `flink-tests` because we use things from `flink-runtime`.

Comment (tillrohrmann) [reply flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusReporterTest.java:88]: Alternatively, this might indicate that the reporters are not super easy to test since we require the registry.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystem.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    InetSocketAddress getInetSocketAddressFromRpcUrl(String url) throws Exception;
+
+    /**
+     * Returns the maximum number of bytes that an RPC message may carry according to the given
+     * configuration. If no limit exists then {@link Long#MAX_VALUE} should be returned.
+     *
+     * @param config Flink configuration
+     * @return maximum number of bytes that an RPC message may carry
+     */
+    long getMaximumMessageSizeInBytes(Configuration config);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystem.java:93]: We might think about creating a separate `RpcSystemUtils` interface for these methods which the `RpcSystem` extends. That way, we have a better interface segregation. Especially, for all the methods which don't need the builders.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystemUtils.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+/** Utils related to the {@link RpcSystem}. */
+public final class RpcSystemUtils {
+
+    /**
+     * Convenient shortcut for constructing a remote RPC Service that takes care of checking for
+     * null and empty optionals.
+     *
+     * @see RpcSystem#remoteServiceBuilder(Configuration, String, String)
+     */
+    public static RpcService createRemoteRpcService(
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystemUtils.java:35]: Does it make sense to move this method into the `RpcUtils` class? The downside of introducing a new `RpcSystemUtils` class is that helper methods are harder to find.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSystem.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+            Configuration configuration,
+            @Nullable String externalAddress,
+            String externalPortRange) {
+        return AkkaRpcServiceUtils.remoteServiceBuilder(
+                configuration, externalAddress, externalPortRange);
+    }
+
+    @Override
+    public InetSocketAddress getInetSocketAddressFromRpcUrl(String url) throws Exception {
+        return AkkaUtils.getInetSocketAddressFromAkkaURL(url);
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSystem.java:49]: Should we unify `AkkaUtils` and `AkkaRpcServiceUtils`?

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @zentol. The changes look good to me. I had a couple of minor comments. +1 for merging after resolving them.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystem.java (hunk 4 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystem.java:93]: 👍

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSystem.java (hunk 6 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcSystem.java:49]: I think it would make sense because as is I couldn't tell why something goes into one or the other. But I would do that in a follow-up after this feature is completed to safe some time on rebasing.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystemUtils.java (hunk 5 above)
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/rpc/RpcSystemUtils.java:35]: yes that would make sense; this separation doesn't really provide a benefit.

File: flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusReporterTest.java (hunk 3 above)
Comment (zentol) [reply flink-metrics/flink-metrics-prometheus/src/test/java/org/apache/flink/metrics/prometheus/PrometheusReporterTest.java:88]: The main issue I think is that we don't have a good MetricGroup for testing, so we have to rely on the runtime implementations which unfortunately rely on the registry. Overall the metric components are too reliant on each other.

I should have one such implementation somewhere, which should allow us to simplify quite a lot of metric-related tests, but I would do that as a follow-up because it implies bigger changes to all reporter tests.
